pkg/controller/controller_utils_test.go

```diff
@@ -209,162 +210,198 @@ func TestControllerExpectations(t *testing.T) {
-	assert.True(t, e.SatisfiedExpectations(logger, rcKey), "Expectations are met but the rc will not sync")
-
-	// Next round of rc sync, old expectations are cleared
-	e.SetExpectations(logger, rcKey, 1, 2)
-	podExp, exists, err = e.GetExpectations(rcKey)
-	assert.NoError(t, err, "Could not get expectations for rc, exists %v and err %v", exists, err)
-	assert.True(t, exists, "Could not get expectations for rc, exists %v and err %v", exists, err)
-	add, del = podExp.GetExpectations()
-
-	assert.Equal(t, int64(1), add, "Unexpected pod expectations %#v", podExp)
-	assert.Equal(t, int64(2), del, "Unexpected pod expectations %#v", podExp)
-
-	// Expectations have expired because of ttl
-	fakeClock.Step(ttl + 1)
-	assert.True(t, e.SatisfiedExpectations(logger, rcKey),
-		"Expectations should have expired but didn't")
+	tests := []struct {
+		name                       string
+		expectationsToSet          []int
+		expectationsExpected       []int64
+		checkExpectationsSatisfied bool
+		expireExpectations         bool
+	}{
+		{
+			name:                       "Expectations have been surpassed",
+			expectationsExpected:       []int64{int64(-1), int64(-1)},
+			checkExpectationsSatisfied: true,
+			expireExpectations:         false,
+		},
+		{
+			name:                       "Old expectations are cleared because of ttl",
+			expectationsToSet:          []int{1, 2},
+			expectationsExpected:       []int64{int64(1), int64(2)},
+			checkExpectationsSatisfied: false,
+			expireExpectations:         true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if len(test.expectationsToSet) > 0 {
+				e.SetExpectations(logger, rcKey, test.expectationsToSet[0], test.expectationsToSet[1])
+			}
+			podExp, exists, err := e.GetExpectations(rcKey)
+			assert.NoError(t, err, "Could not get expectations for rc, exists %v and err %v", exists, err)
+			assert.True(t, exists, "Could not get expectations for rc, exists %v and err %v", exists, err)
+
+			add, del := podExp.GetExpectations()
+			assert.Equal(t, test.expectationsExpected[0], add, "Unexpected pod expectations %#v", podExp)
+			assert.Equal(t, test.expectationsExpected[1], del, "Unexpected pod expectations %#v", podExp)
+			if test.checkExpectationsSatisfied {
+				assert.True(t, e.SatisfiedExpectations(logger, rcKey), "Expectations are met but the rc will not sync")
+			}
+
+			if test.expireExpectations {
+				fakeClock.Step(ttl + 1)
+				assert.True(t, e.SatisfiedExpectations(logger, rcKey), "Expectations should have expired but didn't")
+			}
+		})
+	}
 }
 
 func TestUIDExpectations(t *testing.T) {
 	logger, _ := ktesting.NewTestContext(t)
 	uidExp := NewUIDTrackingControllerExpectations(NewControllerExpectations())
-	rcList := []*v1.ReplicationController{
-		newReplicationController(2),
-		newReplicationController(1),
-		newReplicationController(0),
-		newReplicationController(5),
+	type test struct {
```

Comment:
IIUC `name` and `numReplicas` fully characterize each test case. Thus, could drop `rcKey` and `rcPodNames` from test parameters and then inline `prepareTests` inside `t.Run(test.name, func(t *testing.T) {`?

Reply:
Sure! 👍🏻